src/Microsoft.Identity.Client/Mats/Internal/ActionStore.cs

```diff
@@ -0,0 +1,336 @@
+        }
+
+        public void EndAdalAction(AdalAction action, AuthOutcome outcome, ErrorSource errorSource, string error, string errorDescription)
+        {
+            EndGenericAction(action.ActionId, MatsConverter.AsString(outcome), errorSource, error, errorDescription, string.Empty);
+        }
+
+        public void EndCustomInteractiveActionWithCancellation(CustomInteractiveAction action) => throw new NotImplementedException();
+        public void EndCustomInteractiveActionWithFailure(CustomInteractiveAction action, ErrorSource errorSource, string error, string errorDescription) => throw new NotImplementedException();
+        public void EndCustomInteractiveActionWithSuccess(CustomInteractiveAction action) => throw new NotImplementedException();
+        public void EndInteractiveMsaActionWithCancellation(InteractiveMsaAction action, string accountCid) => throw new NotImplementedException();
+        public void EndInteractiveMsaActionWithFailure(InteractiveMsaAction action, ErrorSource errorSource, string error, string errorDescription, string accountCid) => throw new NotImplementedException();
+        public void EndInteractiveMsaActionWithSignin(InteractiveMsaAction action, string accountCid) => throw new NotImplementedException();
+        public void EndNonInteractiveMsaActionWithFailure(NonInteractiveMsaAction action, ErrorSource errorSource, string error, string errorDescription, string accountCid) => throw new NotImplementedException();
+        public void EndNonInteractiveMsaActionWithTokenRetrieval(NonInteractiveMsaAction action, string accountCid) => throw new NotImplementedException();
+        public void EndWamActionWithCancellation(WamAction action, string wamTelemetryBatch) => throw new NotImplementedException();
+        public void EndWamActionWithFailure(WamAction action, ErrorSource errorSource, string error, string errorDescription, string accountId, string tenantId, string wamTelemetryBatch) => throw new NotImplementedException();
+        public void EndWamActionWithSuccess(WamAction action, string accountId, string tenantId, string wamTelemetryBatch) => throw new NotImplementedException();
+        public IEnumerable<IPropertyBag> GetEventsForUpload() => throw new NotImplementedException();
+
+        public void ProcessAdalTelemetryBlob(IDictionary<string, string> blob)
+        {
+            if (!blob.TryGetValue(AdalTelemetryBlobEventNames.AdalCorrelationIdConstStrKey, out string correlationId))
+            {
+                _errorStore.ReportError("No correlation ID found in telemetry blob", ErrorType.Action, ErrorSeverity.Warning);
+                return;
+            }
+
+            var propertyBags = GetAdalPropertyBagsForCorrelationId(correlationId);
+            int numberOfPropertyBagsMatchingCorrelationId = propertyBags.Count;
+
+            switch (numberOfPropertyBagsMatchingCorrelationId)
+            {
+            case 0:
+                //[vadugar] ToDo: Is this an error?

```

Comment:
Is this class copied from somewhere? It's pretty huge, if it's copied I'd much rather skip reviewing...

Reply:
Everything in the Mats/ directory is a direct port of code from [https://authtelemetry.visualstudio.com/Microsoft%20Auth%20Telemetry%20System/_git/MATS](https://authtelemetry.visualstudio.com/Microsoft%20Auth%20Telemetry%20System/_git/MATS).  For the moment, it's meant to be a class-for-class copy of what's in MATS to ensure functional consistency and get some tests in place.  There will then be a second pass through this to refactor it and remove some of the class complexity which arises due to the original MATS code being DJINNI / C++ for multiple languages/platforms.  So there are a lot of layers in the code (e.g. things like IActionHandle, MatsAction, MsaAction) that we don't/won't need once we design and finalize the MSAL MATS operations.

---
In reply to: [264435112](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/955#discussion_r264435112) [](ancestors = 264435112)

Reply:
Ah ok, this helps!